spacy/pipeline/spancat.py

```diff
@@ -0,0 +1,402 @@
+                    row = spans_index[key]
+                    k = label_map[gold_span.label_]
+                    target[row, k] = 1.0
+            # The target is a flat array for all docs. Track the position
+            # we're at within the flat array.
+            offset += spans.lengths[i]
+        target = self.model.ops.asarray(target, dtype="f")
+        # The target will have the values 0 (for untrue predictions) or 1
+        # (for true predictions).
+        # The scores should be in the range [0, 1].
+        # If the prediction is 0.9 and it's true, the gradient
+        # will be -0.1 (0.9 - 1.0).
+        # If the prediction is 0.9 and it's false, the gradient will be
+        # 0.9 (0.9 - 0.0)
+        d_scores = scores - target
+        loss = float((d_scores ** 2).sum())
+        return loss, d_scores
+
+    def initialize(
+        self,
+        get_examples: Callable[[], Iterable[Example]],
+        *,
+        nlp: Language = None,
+        labels: Optional[Dict] = None,
+    ) -> None:
+        """Initialize the pipe for training, using a representative set
+        of data examples.
+
+        get_examples (Callable[[], Iterable[Example]]): Function that
+            returns a representative sample of gold-standard Example objects.
+        nlp (Language): The current nlp object the component is part of.
+        labels: The labels to add to the component, typically generated by the
+            `init labels` command. If no labels are provided, the get_examples
+            callback is used to extract the labels from the data.
+
+        DOCS: https://spacy.io/api/spancategorizer#initialize
+        """
+        subbatch = []
+        if labels is not None:
+            for label in labels:
+                self.add_label(label)
+        for eg in get_examples():
+            if labels is None:
+                for span in eg.reference.spans[self.key]:
+                    self.add_label(span.label_)
+            if len(subbatch) < 10:
+                subbatch.append(eg)
+        self._require_labels()
+        if subbatch:
+            docs = [eg.x for eg in subbatch]
+            spans = self.suggester(docs)
+            Y = self.model.ops.alloc2f(spans.dataXd.shape[0], len(self.labels))
+            self.model.initialize(X=(docs, spans), Y=Y)
+        else:
+            self.model.initialize()
+
+    def score(self, examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
+        """Score a batch of examples.
+
+        examples (Iterable[Example]): The examples to score.
+        RETURNS (Dict[str, Any]): The scores, produced by Scorer.score_cats.
+
+        DOCS: https://spacy.io/api/spancategorizer#score
+        """
+        validate_examples(examples, "SpanCategorizer.score")
+        self._validate_categories(examples)
+        kwargs = dict(kwargs)
+        kwargs.setdefault("attr", self.key)
+        kwargs.setdefault("attr_name", "spans")  # ensure the dict contains 'spans_f' etc
```

Comment:
As a note after running the example project and look at the `spacy evaluate` results, I don't think using `"spans"` here is going to scale well. We can do things like add `spans_p/r/f` to what's shown in `spacy evaluate`, but you're going to run into problems as soon as you want to evaluate two `spancat` components, which I think might be a pretty common use case?

I'm not entirely sure what to do about the `spacy evaluate` view (possibly just parse the returned scores more flexibly rather than having a fixed list), but I think it'd be better not to use `spans` for everything, and we'd have to figure out how to handle the `default_score_weights`, too, I guess?

I think the current defaults are fine, but it's going to be confusing for users who change the key and have to adjust the score weights by hand. I'm not sure what the options are there...

Reply:
Yea I pretty much had the same internal dialogue when implementing this. 

I kind of assumed that people wouldn't be training two `spancat` components at the same time. I'm trying to imagine the use-case. Obviously you can have a set of different labels in the same `doc.spans[key]` - what would be the use-case of annotating two distinct sets of spans (each with a distinct set of labels?) on the same texts?

> I think the current defaults are fine, but it's going to be confusing for users who change the key and have to adjust the score weights by hand.

This was my reasoning to add this `attr_name` setting to ensure that users don't have to change the score weights, and that the `default_score_weights` just work, for any value of `spans_key`.

Reply:
I think if the default config works for one component that is basically good enough, and as far as I can tell, it would work for `"spans"`? Users with more complicated cases would have to manually adjust the score weights for now, but they'd also have to manually adjust the component name and spans key, too, so it doesn't seem too difficult?

I don't think you'd have two components with the same key, but I think it could be common to have two separate `spancat` components that have separate data + evaluations for two different tasks? If you do it this way, the second component clobbers the scores of the first one during the evaluation in the returned score dict, so you can't even see the scores of the first component.

I think we should not override `attr_name` here because otherwise there's not even a good workaround for evaluation two separate `spancat` components.

Reply:
I'm not sure we're 100% on the same page, so just to elaborate. With the current implementation, if you change the `spans_key` to something else, like `"labeled_spans"`, the `kwargs.setdefault("attr_name", "spans")` actually ensures that works out-of-the box with the default score settings. The defaults are `"spans_r"` etc but before I made this edit, the score dictionary would have put `"labeled_spans_r"` and then the training wouldn't find the required accuracy setting, and it would show `0.0` by default for `"spans_r"`. So the `attr_name` thing is just to normalize the output column name in the final dictionary, to allow changing the key without further edits.

But yes, that approach wouldn't work for multiple `spancat` components, and yes, they should be able to manage the `kwargs` in those cases.

Reply:
Ah, I see what you were thinking with the use of `kwargs`. But I think it wouldn't work during training, where you can't easily pass in those args, or in `spacy evaluate`. I think (at least initially until we have a better option?) having to adjust the score weights names in the config is a better option, so you can train and evaluate two components at the same time.

Reply:
I think `spans_{self.key}` might be a better default, so that the fact that this is an evaluation related to `doc.spans` is always clear?

Reply:
But then the column name during training will be huge... :/

Reply:
Well, we can name the default key `morphologizer_loss` and all will be well.

My actual suggestion would be use a really short default key that's not `spans` (which is going to obviously be confusing, just like you said in the project), so something like `sc`. Then the default scores are `spans_sc_p/r/f` and I think that's manageable for the output? With multiple components you could have `spans_sc1_f` and `spans_sc1_f` and I think it would still be readable even with the truncated column headers?

I think then we could just provide `spans_{self.key}` as `attr` and not need to make any changes to the scoring method, given the custom `getattr` that manages the rest? I haven't tried that out yet, though.

Reply:
Okay, the `default_score_weights` are a bit more of a hassle than I thought, since there's no way to parameterize them, and I think the `[training.score_weights]` overrides weren't all being added as expected. Here's what would be required for what I described above:

https://github.com/adrianeboyd/spaCy/commit/79a0403b02c73f4c30d7d40ab2e71234a69efe5f

And the output with `sc` and `sc2` looks like this:

```none
E    #       LOSS SPANCAT  LOSS SPANCAT2  SPANS_SC_F  SPANS_SC_P  SPANS_SC_R  SPANS_SC2_F  SPANS_SC2_P  SPANS_SC2_R  SCORE 
---  ------  ------------  -------------  ----------  ----------  ----------  -----------  -----------  -----------  ------
  0       0        184.50         184.50        2.31        1.50        4.98         2.73         1.90         4.83    0.03
  0     200       2658.05        2696.12        0.00        0.00        0.00         0.00         0.00         0.00    0.00
  0     400       1309.46        1350.80       30.41       83.67       18.58        33.45        74.09        21.60    0.32
```

(For the future: these two components that are being trained on the exact same data but the results aren't identical, so something may be non-deterministic in the model somewhere.)

Reply:
This suggestion is fine by me, and yea I would like to have a different default key than "spans". "sc" is fine I guess. If we want to support two spancat components in the same pipeline, we'll need to do something like this. Do you want to commit?

Reply:
I do think it's best to make sure the first version of this can handle separate scoring for multiple components.

Ideally, it'd be nice to parameterize the default score weights, but I don't think that's realistic for v3.1, so the `overrides` change looks like the easiest alternative.

However, I want to make sure I understand the side effects of the `overrides` issue. It's possible it will make the train output really ugly if you're just updating one component in a long pipeline and you'd end up with a lot of extra values shown in the table that wouldn't be there otherwise, but I need to test it out.

Reply:
It looks like the scores for the frozen components are included in the output either way, so I don't think allowing the extra score overrides makes much of a UX difference. And this change would only potentially add some extra scores/columns to the output, but I really don't think it has the potential to break any existing functionality.